src/audio/kpb.c

```diff
@@ -2060,6 +2063,50 @@ static inline void kpb_change_state(struct comp_data *kpb,
 	kpb->state_log = (kpb->state_log << 4) | state;
 }
 
+static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+			     int max_data_size)
+{
+	const struct kpb_micselector_config *mic_sel = data;
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	const size_t mic_cnt = kpb->config.channels - KPB_REFERENCE_SUPPORT_CHANNELS;
+	const uint8_t valid_mask = KPB_COUNT_TO_BITMASK(mic_cnt);
+	size_t i;
+
+	if ((valid_mask & mic_sel->mask) == 0) {
+		comp_err(dev, "error: invalid micselector bit mask");
+		return -EINVAL;
+	}
+	 /* selected mics counter */
+	size_t num_of_sel_mic = 0;
+
+	for (i = 0; i < mic_cnt; i++) {
+		if (KPB_IS_BIT_SET(mic_sel->mask, i)) {
+			kpb->offsets[num_of_sel_mic] = i;
+			num_of_sel_mic++;
+		}
+	}
+	kpb->num_of_sel_mic = num_of_sel_mic;
+	kpb->num_of_in_channels = kpb->config.channels;
+	kpb->mic_sel.mask = mic_sel->mask;
+	return 0;
+}
+
+static int kpb_set_large_config(struct comp_dev *dev, uint32_t param_id,
+				bool first_block,
+				bool last_block,
+				uint32_t data_offset,
+				const char *data)
+{
+	comp_info(dev, "kpb_set_large_config()");
+
+	switch (param_id) {
+	case KP_BUF_CLIENT_MIC_SELECT:
```

Comment:
We don't at the moment support for other than byte control type in IPC4, so e.g. enum is not possible. But please align with @ujfalusi how to utilize the param_id bits space in this set large config IPC so we are later able to add the other control types, and several controls of same type for a component.

Reply:
@singalsu, @fkwasowi, I think this is not an enum type of control but a binary. It is expecting some mask sent down, no?

So, I think LARGE_CONFIG, param_id and the payload is u32 is fine here.

We are mixing up things with `enum` and **enum**. The param_id has to be unique to target a certain configuration, defining that with enum is fine, could be #define as well, but we like enums.

The enum type or for the matter: multiple, same type controls/parameters for a same module is the issue at the end of the debate.

But if we look a bit further: if a module have more than one parameter of the same type (volume, switch, enum, bytes), how that should be handled? If we need to add another mic select  for this module for example. I guess we will have a new param_id?
Since the binary is already assuming this (we have param_id + payalod), I believe this is going to be the case going forward?
So, each parameter (like multiple volume for a single module) is differentiated based on the param_id?

Reply:
It seems like value -1 means all channels, while values 0, 1, .. refer to channels. So setting left and right channel to a different gain value would need two set large config messages, right?